lrpar/src/lib/ctbuilder.rs

```diff
@@ -374,24 +383,59 @@ where
         }
 
         let inc = read_to_string(grmp).unwrap();
-        let grm = YaccGrammar::<StorageT>::new_with_storaget(yk, &inc)
-            .map_err(|errs| {
-                errs.iter()
-                    .map(|e| {
-                        let mut line_cache = NewlineCache::new();
-                        line_cache.feed(&inc);
-                        if let Some((line, column)) =
-                            line_cache.byte_to_line_num_and_col_num(&inc, e.spans()[0].start())
-                        {
-                            format!("{} at line {line} column {column}", e)
+        let ast_validation = ASTWithValidityInfo::new(yk, &inc);
+        let warnings = ast_validation.ast().warnings();
+        let spanned_fmt = |x: &dyn Spanned, inc: &str, line_cache: &NewlineCache| {
+            if let Some((line, column)) =
+                line_cache.byte_to_line_num_and_col_num(inc, x.spans()[0].start())
+            {
+                format!("{} at line {line} column {column}", x)
+            } else {
+                format!("{}", x)
+            }
+        };
+
+        let res = YaccGrammar::<StorageT>::new_from_ast_with_validity_info(yk, ast_validation);
+        let grm = match res {
+            Ok(_) if self.warnings_are_errors && !warnings.is_empty() => {
+                let mut line_cache = NewlineCache::new();
+                line_cache.feed(&inc);
+                return Err(ErrorString(
+                    warnings
+                        .iter()
+                        .map(|w| spanned_fmt(w, &inc, &line_cache))
+                        .collect::<Vec<_>>()
+                        .join("\n"),
+                ))?;
+            }
+            Ok(grm) => {
+                if !warnings.is_empty() {
+                    let mut line_cache = NewlineCache::new();
+                    line_cache.feed(&inc);
+                    for w in warnings {
+                        // Assume if this variable is set we are running under cargo.
+                        if std::env::var("OUT_DIR").is_ok() {
+                            println!("cargo:warning={}", spanned_fmt(&w, &inc, &line_cache));
                         } else {
-                            format!("{}", e)
+                            eprintln!("{}", spanned_fmt(&w, &inc, &line_cache));
                         }
-                    })
-                    .collect::<Vec<_>>()
-                    .join("\n")
-            })
-            .map_err(ErrorString)?;
+                    }
+                }
+                grm
+            }
+            Err(errs) => {
+                let mut line_cache = NewlineCache::new();
+                line_cache.feed(&inc);
+                return Err(ErrorString(
```

Comment:
I just realised that I don't know what `ErrorString` is! Any clues?

Reply:
It's just a `Debug` impl for string which uses `Display` rather than `Debug`, so that it doesn't emit `"`, at the beginning/end of the string or "\n" literally, at the end of each line, etc.

Reply:
Oh, I see it's defined here https://github.com/softdevteam/grmtools/blob/4311b961a1ef081c7c3136786a5e3108ebc0e851/lrpar/src/lib/ctbuilder.rs#L94.